src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Ricci.cpp

```diff
@@ -0,0 +1,163 @@
+  if (UNLIKELY(get_size(get<0, 0>(*ricci)) != get_size(get<0, 0, 0>(phi)))) {
+    *ricci = tnsr::ii<DataType, VolumeDim, Frame>(get_size(get<0, 0, 0>(phi)));
+  }
+
+  TempBuffer<tmpl::list<::Tags::TempIjj<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempijk<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempi<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempi<1, VolumeDim, Frame, DataType>,
+                        ::Tags::TempI<0, VolumeDim, Frame, DataType>,
+                        ::Tags::TempI<1, VolumeDim, Frame, DataType>>>
+      local_buffer(get_size(get<0, 0>(inverse_spatial_metric)));
+
+  // New variable to avoid recomputing in nested loops
+  auto& phi_uplolo =
+      get<::Tags::TempIjj<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = i; j < VolumeDim; ++j) {  // Symmetry
+        phi_uplolo.get(k, i, j) = 0.;
+        for (size_t l = 0; l < VolumeDim; ++l) {
+          phi_uplolo.get(k, i, j) +=
+              0.5 * inverse_spatial_metric.get(k, l) * phi.get(l, 1 + i, 1 + j);
+        }
+      }
+    }
+  }
+
+  // New variable to avoid recomputing in nested loops
+  auto& phi_loloup =
+      get<::Tags::Tempijk<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = 0; j < VolumeDim; ++j) {
+        phi_loloup.get(i, j, k) = 0.;
+        for (size_t l = 0; l < VolumeDim; ++l) {
+          phi_loloup.get(i, j, k) +=
+              0.5 * inverse_spatial_metric.get(k, l) * phi.get(i, 1 + j, 1 + l);
+        }
+      }
+    }
+  }
+
+  // New variable to avoid recomputing in nested loops
+  auto& phi_trace1 =
+      get<::Tags::Tempi<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t i = 0; i < VolumeDim; ++i) {
+    phi_trace1.get(i) = 0.;
+    for (size_t j = 0; j < VolumeDim; ++j) {
+      phi_trace1.get(i) += phi_loloup.get(i, j, j);
+    }
+  }
+  auto& phi_trace2 =
+      get<::Tags::Tempi<1, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t i = 0; i < VolumeDim; ++i) {
+    phi_trace2.get(i) = 0.;
+    for (size_t j = 0; j < VolumeDim; ++j) {
+      phi_trace2.get(i) += phi_uplolo.get(j, j, i);
+    }
+  }
+
+  auto& phi_trace1_up =
+      get<::Tags::TempI<0, VolumeDim, Frame, DataType>>(local_buffer);
+  auto& phi_trace2_up =
+      get<::Tags::TempI<1, VolumeDim, Frame, DataType>>(local_buffer);
+
+  raise_or_lower_index(make_not_null(&phi_trace1_up), phi_trace1,
+                       inverse_spatial_metric);
+  raise_or_lower_index(make_not_null(&phi_trace2_up), phi_trace2,
+                       inverse_spatial_metric);
+
+  for (size_t i = 0; i < VolumeDim; ++i) {
+    for (size_t j = i; j < VolumeDim; ++j) {  // Symmetry
+      ricci->get(i, j) = 0.;
+      for (size_t p = 0; p < VolumeDim; ++p) {
+        ricci->get(i, j) +=
```

Comment:
I haven to double-checked that this formula agrees with the formula in your help text. But I'm confused because your helptext uses spacetime indices, but here you're just using spatial indices

Reply:
Redid the help text, hope that clarifies whats coded up.